File: app/services/team_service.py
from collections.abc import Sequence

from app.models.team import Team, TeamMember
from app.repositories.team_repo import TeamRepository
from app.schemas.team import TeamCreate, TeamMembersSet, TeamUpdate
from app.services.exceptions import (
    BadRequestError,
    ConflictError,
    NotFoundError,
    UnprocessableEntityError,
)


class TeamService:
    def __init__(self, team_repo: TeamRepository) -> None:
        self.team_repo = team_repo

    async def list_teams(self) -> Sequence[Team]:
        return await self.team_repo.list_teams()

    async def get_team_by_id(self, team_id: int) -> Team:
        team = await self.team_repo.get_team_by_id(team_id)
        if team is None:
            raise NotFoundError("Team not found")
        return team

    async def _ensure_name_available(self, name: str, current_team_id: int | None = None) -> None:
        existing = await self.team_repo.get_team_by_name(name)
        if existing is not None and existing.id != current_team_id:
            raise ConflictError("Team name already taken")

    async def create_team(self, data: TeamCreate) -> Team:
        await self._ensure_name_available(data.name)
        return await self.team_repo.create_team(data.name, data.description)

    async def update_team(self, team_id: int, data: TeamUpdate) -> Team:
        team = await self.get_team_by_id(team_id)
        fields = data.model_dump(exclude_unset=True)

        if "name" in fields:
            await self._ensure_name_available(fields["name"], current_team_id=team.id)

        if not fields:
            return team

        return await self.team_repo.update_team(team, fields)

    async def set_members(self, team_id: int, data: TeamMembersSet) -> Team:
        team = await self.get_team_by_id(team_id)

        if data.lead_id is not None and data.lead_id not in data.user_ids:
            raise BadRequestError("Lead must be a team member")

        users = await self.team_repo.get_users_by_ids(data.user_ids)
        found_user_ids = {user.id for user in users}
        missing_user_ids = sorted(set(data.user_ids) - found_user_ids)
        if missing_user_ids:
            raise NotFoundError(f"Users not found: {missing_user_ids}")

        desired_user_ids = set(data.user_ids)
        current_memberships = await self.team_repo.get_memberships_for_team(team.id)
        current_by_user_id = {membership.user_id: membership for membership in current_memberships}

        for membership in current_memberships:
            if membership.user_id not in desired_user_ids:
                await self.team_repo.delete_membership(membership)

        incoming_memberships = await self.team_repo.get_memberships_for_users(data.user_ids)
        incoming_by_user_id = {
            membership.user_id: membership for membership in incoming_memberships
        }

        for user_id in data.user_ids:
            membership = incoming_by_user_id.get(user_id)
            if membership is not None and membership.team_id == team.id:
                continue

            if membership is not None:
                await self._move_user_from_previous_team(membership)

            if user_id not in current_by_user_id or (
                membership is not None and membership.team_id != team.id
            ):
                await self.team_repo.add_membership(team.id, user_id)

        await self.team_repo.set_team_lead(team, data.lead_id)
        refreshed_team = await self.team_repo.get_team_by_id(team.id)
        if refreshed_team is None:
            raise UnprocessableEntityError("Team update failed")
        return refreshed_team

    async def _move_user_from_previous_team(self, membership: TeamMember) -> None:
        previous_team = membership.team
        if previous_team.lead_id == membership.user_id:
            await self.team_repo.set_team_lead(previous_team, None)
        await self.team_repo.delete_membership(membership)

File: backend/src/main/java/com/skilltracker/service/TeamService.java
package com.skilltracker.service;

import com.skilltracker.domain.Team;
import com.skilltracker.domain.TeamMember;
import com.skilltracker.dto.TeamCreateRequest;
import com.skilltracker.dto.TeamMembersSetRequest;
import com.skilltracker.dto.TeamResponse;
import com.skilltracker.dto.TeamUpdateRequest;
import com.skilltracker.exception.BadRequestException;
import com.skilltracker.exception.ConflictException;
import com.skilltracker.exception.NotFoundException;
import com.skilltracker.repository.TeamMemberRepository;
import com.skilltracker.repository.TeamRepository;
import com.skilltracker.repository.UserRepository;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TeamService {

    private final TeamRepository teamRepository;
    private final TeamMemberRepository teamMemberRepository;
    private final UserRepository userRepository;
    private final TeamMapper teamMapper;
    private final ReadBack readBack;

    public TeamService(
            TeamRepository teamRepository,
            TeamMemberRepository teamMemberRepository,
            UserRepository userRepository,
            TeamMapper teamMapper,
            ReadBack readBack) {
        this.teamRepository = teamRepository;
        this.teamMemberRepository = teamMemberRepository;
        this.userRepository = userRepository;
        this.teamMapper = teamMapper;
        this.readBack = readBack;
    }

    @Transactional(readOnly = true)
    public List<TeamResponse> listTeams() {
        return teamMapper.toResponses(teamRepository.findAllByOrderByIdAsc());
    }

    @Transactional(readOnly = true)
    public TeamResponse getTeam(Integer teamId) {
        return teamMapper.toResponse(requireTeam(teamId));
    }

    @Transactional
    public TeamResponse createTeam(TeamCreateRequest request) {
        ensureNameAvailable(request.name(), null);

        Team team = teamRepository.save(new Team(request.name(), request.description()));
        readBack.flushAndDetach();
        return teamMapper.toResponse(requireTeam(team.getId()));
    }

    @Transactional
    public TeamResponse updateTeam(Integer teamId, TeamUpdateRequest request) {
        Team team = requireTeam(teamId);

        if (request.name().isPresent() && request.name().value() != null) {
            ensureNameAvailable(request.name().value(), team.getId());
            team.setName(request.name().value());
        }
        if (request.description().isPresent()) {
            team.setDescription(request.description().value());
        }

        teamRepository.save(team);
        readBack.flushAndDetach();
        return teamMapper.toResponse(requireTeam(teamId));
    }

    /**
     * Replaces the membership list wholesale. A user belongs to at most one team, so anyone joining
     * from another team is removed there first, and that team loses its lead if it was that user.
     */
    @Transactional
    public TeamResponse setMembers(Integer teamId, TeamMembersSetRequest request) {
        Team team = requireTeam(teamId);
        List<Integer> userIds = request.userIds();

        if (request.leadId() != null && !userIds.contains(request.leadId())) {
            throw new BadRequestException("Lead must be a team member");
        }
        requireUsersExist(userIds);

        Set<Integer> desired = new HashSet<>(userIds);
        List<TeamMember> currentMemberships = teamMemberRepository.findByTeamIdOrderById(teamId);
        Set<Integer> currentUserIds =
                currentMemberships.stream().map(TeamMember::getUserId).collect(Collectors.toSet());

        for (TeamMember membership : currentMemberships) {
            if (!desired.contains(membership.getUserId())) {
                teamMemberRepository.delete(membership);
            }
        }
        readBack.flushAndDetach();

        Map<Integer, TeamMember> incoming = new HashMap<>();
        if (!userIds.isEmpty()) {
            teamMemberRepository
                    .findByUserIdIn(userIds)
                    .forEach(membership -> incoming.put(membership.getUserId(), membership));
        }

        for (Integer userId : userIds) {
            TeamMember membership = incoming.get(userId);
            if (membership != null && membership.getTeamId().equals(teamId)) {
                continue;
            }
            if (membership != null) {
                moveUserFromPreviousTeam(membership);
            }
            if (!currentUserIds.contains(userId) || membership != null) {
                teamMemberRepository.save(new TeamMember(teamId, userId));
            }
        }

        Team reloaded = requireTeam(teamId);
        reloaded.setLeadId(request.leadId());
        teamRepository.save(reloaded);
        readBack.flushAndDetach();

        return teamMapper.toResponse(requireTeam(teamId));
    }

    private void moveUserFromPreviousTeam(TeamMember membership) {
        Team previousTeam = requireTeam(membership.getTeamId());
        if (membership.getUserId().equals(previousTeam.getLeadId())) {
            previousTeam.setLeadId(null);
            teamRepository.save(previousTeam);
        }
        teamMemberRepository.delete(membership);
        readBack.flushAndDetach();
    }

    private void requireUsersExist(List<Integer> userIds) {
        if (userIds.isEmpty()) {
            return;
        }

        Set<Integer> found = userRepository.findAllByIdIn(userIds).stream()
                .map(user -> user.getId())
                .collect(Collectors.toSet());
        Set<Integer> missing = new TreeSet<>(userIds);
        missing.removeAll(found);
        if (!missing.isEmpty()) {
            throw new NotFoundException("Users not found: " + new ArrayList<>(missing));
        }
    }

    private void ensureNameAvailable(String name, Integer currentTeamId) {
        teamRepository
                .findByNameIgnoreCase(name)
                .filter(existing -> !existing.getId().equals(currentTeamId))
                .ifPresent(existing -> {
                    throw new ConflictException("Team name already taken");
                });
    }

    private Team requireTeam(Integer teamId) {
        return teamRepository.findById(teamId).orElseThrow(() -> new NotFoundException("Team not found"));
    }
}
